---
title: "Haider Data Figures"
author: "Fred White"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  bookdown::html_document2:
    toc: true
    theme: united
    toc_depth: 3
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F, echo = F)
```



```{r , echo=FALSE, include=FALSE}
source("DATA_SIM_FUNCS.R")
```

```{r, echo = FALSE, include=FALSE}
# getwd()
library(devtools)
library(ggplot2)
library(gASCA)
library(patchwork)
library(readxl)
library(ggfortify)
library(ggrepel)
library(grid)
library(gridExtra)
library(tidyr)
library(tidyverse)
library(reshape2)
library(scales)
library(matrixStats)
library(DESeq2)
library(MASS)
library(glmnet)
library(doParallel)    
library(pheatmap)
library(data.table)
library(plotly)
library(pls)
library(pracma)

```



```{r}
Date <- Sys.Date()

species <- "Oryza sativa"
# ncores = 16
# registerDoParallel(cores = ncores)

baits <- c("Os11g0587000","Os04g0550600","Os01g0746400","Os01g0700900","Os06g0565100","Os01g0700500","Os08g0547900")


sls <- c("Os01t0700300-00", "Os01t0700900-01", "Os01t0746400-01", "Os03t0263300-01", "Os03t0408600-01", "Os02t0817900-01", "Os04t0550600-01", "Os11t0587000-01", "Os12t0405200-01")
URPS <- c("Os08t0299400-01","Os03t0406100-01","Os10t0116800-01","Os11t0658900-01","Os02t0609000-01","Os11t0151700-01","Os08t0564000-01","Os08t0280100-01","Os05t0542150-01")


ref <- read.table("Rice_Annotation/Oryza_sativa.IRGSP-1.0.50.gtf", sep = "\t")      
raw_counts <- read.table("Data_Haider/raw_counts.tsv", sep = "\t", header = T)

meta <- readRDS("Data_Haider/RICE_META.rds")
```



```{r}


ref <- ref[which(ref$V3 == "transcript"),]
ref$ID <- gsub(".*gene_id ","",ref$V9)
ref$ID <- gsub(";.*","",ref$ID)

ref$transcript <- gsub(".*transcript_id ","",ref$V9)
ref$transcript <- gsub(";.*","",ref$transcript)

ref$biotype <- gsub(".*transcript_biotype ","",ref$V9)
ref$biotype <- gsub(";.*","",ref$biotype)


baits_transcript <- ref[which(ref$ID %in% baits),c(10:11)]
```


```{r}
meta <- meta[match(colnames(raw_counts[,-c(1:6)]),rownames(meta)),]

meta$growth_condition[which(meta$growth_condition == "-P")] <- "P-"
meta$growth_condition[which(meta$growth_condition == "+P")] <- "P+"

```


```{r}
#filtering low/no expression genes
counts <- raw_counts


colnames(counts)[-c(1:6)] <- meta$ID

conds <- unique(gsub("_1_.*","_",colnames(counts)[-c(1:6)]))


####filter per experimental condition on number of zeroes
i <- NULL
cond_zeroes <- data.frame(matrix())
for(i in 1:length(conds)){
  
  #print(grep(conds[i], colnames(samples)))
  
  zeroes <- rowSums(counts[,grep(conds[i], colnames(counts), fixed = T)] == 0)
  cond_zeroes <- cbind(cond_zeroes,zeroes)
  
  
}

cond_zeroes <- cond_zeroes[,-1]
colnames(cond_zeroes) <- conds


counts2 <- counts[rowSums(cond_zeroes==0) >= 3,]


cond_zeroes <- cond_zeroes[rowSums(cond_zeroes==0) >= 3,]


counts3 <- counts2[which(rowMax(data.matrix(counts2[,-c(1:6)])) < 5000),]

```



```{r}
rownames(meta) <- NULL




dds <- DESeqDataSetFromMatrix(countData = counts3[,-c(1:6)], 
                              colData = meta, 
                              design = ~ ID) 


vsd <- varianceStabilizingTransformation(object = dds, 
                                         blind = TRUE,           # do not take the design formula into account. 
                                         # best practice for sample-level QC
                                         fitType = "parametric")

# extract the matrix of variance stabilised counts
variance_stabilised_counts <- assay(vsd)

saveRDS(vsd, "Data_Haider/Rice_Counts4ASCA.RDS")
saveRDS(meta, "Data_Haider/Rice_Meta4ASCA.RDS")
```






```{r real data import}

X <- readRDS("Data_Haider/Rice_Counts4ASCA.RDS")
meta <- readRDS("Data_Haider/Rice_Meta4ASCA.RDS")
colnames(meta)[3] <- c("time")


baits <- paste0(c("Os11t0587000","Os04t0550600","Os01t0746400","Os01t0700900"),"-01") 
#D27 CCD7 CCD8 CYP711A2
# baits2 <- c("Os01t0700300", "Os01t0701700")  
baits2 <- c("Os01t0700300-00", "Os02t0817900-01")  # identified methyltransferases
```


```{r}
meta$time <- as.character(meta$time)

meta$time[which(meta$time == "8")] <- "1"
meta$time[which(meta$time == "10")] <- "3"
meta$time[which(meta$time == "14")] <- "7"
meta$time[which(meta$time == "15")] <- "8"


X <- t(assay(X))
ref <- cbind.data.frame(Feature = colnames(X), feature = colnames(X), Baits = colnames(X))



REF <- fread("Rice_Annotation/IRGSP-1.0_representative_annotation_2022-09-01.tsv",sep = "\t", header = T)

ref <- ref[which(ref$Feature %in% REF$Transcript_ID),]
ref <- merge(ref,REF[,c("Transcript_ID","Description")], by.x = "Feature", by.y = "Transcript_ID")




```

```{r}
meta <- meta[order(meta$time),]
meta <- meta[order(meta$growth_condition),]
```


```{r}
X <- X[match(meta$ID, rownames(X)),]
```



```{r heatmap}
# h <- readRDS("heat_comp.RDS")
# levels(h[[2]]$time) <- c("1","3","7","8")

# make row annotation

row_ann <- data.frame("Pathway" = rep(c("SL","URP"),each = 9))
rownames(row_ann) <-  c(sls,URPS)         #c(pathway,rownames(ASCA_cands3)[1:9])

ann_colours <- list(g_c = c("P-" = "#F8766D","P+" = "lightblue"),
                    time = c("1" = "#fde725","3" = "#35b779","7" = "#31688e","8" = "#440154"),
                    Pathway = c("SL" = "purple", "URP" = "lightgreen"))

h <- t(X[,c(sls,URPS)])#t(X[,c(pathway,rownames(ASCA_cands3)[1:9])])  #psi_ref[,1]

col_ann <- meta
rownames(col_ann) <- col_ann$ID
col_ann <- col_ann[,-c(1)]
colnames(col_ann)[1] <- "g_c"

heat <- pheatmap::pheatmap(t(scale(t(h))),  #
                             cluster_rows = F,
                             cluster_cols = F,
                             show_rownames = T,
                             show_colnames = F,
                             annotation_col = col_ann,
                           annotation_row = row_ann,
                           annotation_colors = ann_colours,
                           treeheight_row = 10,
                           fontsize = 8,
                           main = "Scaled Gene Expression",silent = T)

tiff("Fig3.tif", res = 150, height = 960, width = 960)   #MASCARA_heatmap.png
grid::grid.newpage()
grid::grid.draw(heat$gtable)
dev.off()


grid::grid.newpage()
grid::grid.draw(heat$gtable)

```



```{r supplementary heatmap id function table}
library(writexl)



supplementary <- REF[which(REF$Transcript_ID %in% c(sls,URPS)),]
supplementary <- as.data.frame(supplementary)
rownames(supplementary) <- supplementary$Transcript_ID
supplementary <- supplementary[c(sls,URPS),]
rownames(supplementary) <- NULL
write_xlsx(supplementary, "S_Table_1_Heatmap_ID_functions.xlsx")


```


```{r}
X <- X[,which(colnames(X) %in% ref$Feature)]
ref <- ref[match(colnames(X),ref$Feature),]




SL_ref <- ref[grep("trigolact",ref$Description),]

SL_ref <- ref[grep("trigolact",ref$Description),]
SL_ref <- SL_ref[-which(SL_ref$Feature %in% c("Os01t0701400-01","Os03t0203200-01","Os01t0763200-01","Os04t0668900-01","Os06t0154200-01","Os11t0104300-01","Os08t0250900-01", "Os03t0408600-01")),]





PSI_ref <-read.table("Rice_Annotation/PSI_genes.txt", sep = "\t", header = T)

GA_ref <- read_xlsx("Rice_Annotation/GA pathway genes-IDs.xlsx")
GA_ref$ID <- gsub("g","t",GA_ref$ID)
GAs <- ref[which(gsub("-.*","",ref$Feature) %in% GA_ref$ID),1]
```


```{r}

ref$Baits[-which(ref$Baits %in% c(sls,URPS))] <- "Other"
ref$Baits[which(ref$Baits %in% sls)] <- "POI"
ref$Baits[which(ref$Baits %in% URPS)] <- "URP"


ref$Baits <- factor(ref$Baits, levels = c("Other","POI","URP"))




```



```{r, echo = FALSE}
#### VST inverse #######
#mean/sd relationship plots
##Power transformation of counts/concentration matrix
m_sd_n <- M_SD(X, meta)

plot_normal <- m_sd_n[[3]]
```



```{r, echo=FALSE, include=FALSE}
# db <- cbind.data.frame(rownames(meta),meta[,c(7,11)],X)

db <- cbind.data.frame(meta,X)


# db[,1:3] <- apply(db[,1:3],2,as.character)
mc <- reshape2::melt(db)
colnames(mc)[4:5] <- c("Feature","Expression")
```

```{r, echo = FALSE}
dist_norm <- ggplot(mc, aes(x = Expression)) + geom_density(alpha = 0.4)+
  scale_y_continuous(labels = percent_format()) + 
  theme_bw()+
  ggtitle("Distribution")
```


```{r, echo=FALSE}
dist_nb <- ggplot(mc, aes(x = Expression)) + geom_density(alpha = 0.4)+
  scale_y_continuous(labels = percent_format()) + 
  scale_x_continuous(guide = guide_axis(check.overlap = TRUE)) +
  theme_bw()+
  ggtitle("NB transformed")


```



```{r, echo = FALSE, include = F}
m <- meta
m[,c("growth_condition","time")] <- lapply(m[,c("growth_condition","time")],factor)


res_ASCAplus_nb <- ASCA_decompose(d = m[,c("growth_condition","time")], x = X,
                               f = "growth_condition + time + growth_condition:time")
```


```{r, echo = FALSE}

 minT <-  res_ASCAplus_nb$decomposition$growth_condition + res_ASCAplus_nb$decomposition$`growth_condition:time`
  PCD <- prcomp(minT)
  
  newdata <- res_ASCAplus_nb$decomposition$growth_condition + res_ASCAplus_nb$decomposition$`growth_condition:time` + res_ASCAplus_nb$residuals
  

  PCDE <- predict(PCD, newdata)

```  


```{r, echo = F, include = FALSE}
a_c.PCD <- PCD
SL <- SL_plot(a_c.PCD, meta)
```



```{r, echo = F, fig.cap="Top: distribution and mean sd relationship. Bottom: scores and loadings of the main effect; growth condition + interaction"}
ve1 <- paste0(round(summary(a_c.PCD)$importance[2,1:2] * 100, digits = 2), "%")

loadingplot <- function(df, meta, ve, ref, ...){
  #df is PCD$rotation

  #rematch REF, ref, df
  df <- as.data.frame(df)
  df <- cbind.data.frame(ref,df)

  ggplot(df, ...) +

    xlab(paste0("PC1 ",ve[1])) +
    ylab(paste0("PC2 ",ve[2])) +

geom_point(data = df, size = 5) +
    scale_x_continuous(guide = guide_axis(check.overlap = TRUE)) +
    guides(shape = guide_legend(override.aes = list(size = 0.5))) +
    ggtitle("Loadings") +
    theme_bw()

}



l <- loadingplot(a_c.PCD$rotation, meta, ve = ve1, ref = ref, aes(x = PC1,y = PC2, colour = Baits, label = Description, alpha = Baits))  + theme(legend.title=element_blank())



l <- l + scale_color_manual(values = c("Other" = "black", "POI" = "purple","URP" = "green")) + coord_fixed() + theme(aspect.ratio = 1,text=element_text(size=18))



meta2 <- meta
colnames(meta2)[2] <- "g_c"

scoreplot <- function(df, meta, ve, ...){
  
  #df <- as.data.frame(df)
  df <- cbind.data.frame(meta,df)
  df$time <- factor(df$time) #, levels = c(1:4)
  df$g_c <- factor(df$g_c)
  
  
  ggplot(df, ...)+
    geom_point(size = 5) +
    xlab(paste0("PC1 ",ve[1])) +
    ylab(paste0("PC2 ",ve[2])) + 
    theme_bw() 
    
  
}
s <- scoreplot(PCDE, meta2, ve1, aes(x= PC1, y = PC2, colour = g_c, shape = time)) + ggtitle("Scores")
s <- s + coord_fixed() + theme(aspect.ratio = 1, text=element_text(size=18))

#+ guide_area()# + plot_layout(design = design)
qq <- qqplot(newdata, meta, aes(x=X_2000, y= X_1999, colour = growth_condition, shape = time))


ASCA_cands <- get_ASCA_cands(a_c.PCD)
colnames(ASCA_cands) <- c("VIP", "PC1","PC2")
#kable(head(ASCA_cands,10))

CAND_TAB <- round(ASCA_cands[1:10,], 4)
CAND_TAB <- cbind.data.frame(rownames(CAND_TAB), CAND_TAB)
colnames(CAND_TAB)[1] <- "Feature"


fill <- rep(c("grey95", "grey90"), nrow(CAND_TAB))
fill[which(CAND_TAB$Feature %in% baits)] <- "yellow"

theme1 <- ttheme_default(core = list(
  fg_params = list(fontface=c(rep("plain", nrow(CAND_TAB)))),
  bg_params = list(fill = fill)),
  base_size = 5, padding = unit(c(2, 2), "mm"))

candtab <- gridExtra::tableGrob(CAND_TAB, rows = NULL, theme = theme1)

# (dist_norm + plot_normal)/(s + l)

```



```{r}
(s + l)

tiff("Fig4.tif", width = 960, height = 480)   #rice_s_l.png
(s + l)
dev.off()
```

```{r}

get_ASCA_cands2 <- function(PCD, meta, distance_calc= FALSE, baits = NULL, spikes = NULL, ret_candN = nrow(PCD$rotation)){
  #############
  
  if(distance_calc== TRUE){
    
    cands <- ranked_dist(baits,PCD)
    
  }else{
    absload <- abs(data.matrix(PCD$rotation[,1:2])) %*% diag(summary(PCD)$importance[2,1:2])
    combscore <- rowSums(absload[,1:2])
    
    
    orderedload <- cbind(combscore, PCD$rotation[,1:2])
    cands <- as.data.frame(orderedload[order(orderedload[,1], decreasing = T),])
    colnames(cands) <- c("VIP", "PC1","PC2")
    
    cands <- round(cands[1:ret_candN,], 4)
    
    
  }
  
  return(cands)
  
  
}

pathway <- ref[which(ref$Baits == "POI"),1]


ASCA_cands2 <- get_ASCA_cands2(a_c.PCD)

ac_top <- ASCA_cands2[1:(nrow(ASCA_cands2)/2),]

ASCA_cands2 <- ASCA_cands2[which(ASCA_cands2[,2] < 0),]
ASCA_cands2 <- ASCA_cands2[order(ASCA_cands2[,2]),]
ASCA_cands2 <- ASCA_cands2[-which(rownames(ASCA_cands2) %in% pathway),]

ASCA_cands3 <- ASCA_cands2[-which(rownames(ASCA_cands2) %in% c("Os06t0570600-00","Os06t0570566-00", "Os09t0321200-00","Os06t0651900-01", "Os10t0444566-00")),]



qqplot <- function(data, meta, ...){

  df <- cbind.data.frame(meta,data)

  df$time <- factor(df$time, levels = unique(meta$time))
  df$growth_condition <- factor(df$growth_condition)


  ggplot(df, ...) +
    geom_point() +
    theme_bw()


}

meta$time <- rep(c(1,3,7,8), each = 3, 2)




PSI_ref <-read.table("Rice_Annotation/PSI_genes.txt", sep = "\t", header = T)
psi_ref <- PSI_ref[which(PSI_ref[,1] %in% colnames(res_ASCAplus_nb$residuals)),]


PSI_cands <- rownames(ASCA_cands[which(rownames(ASCA_cands) %in% psi_ref[,1]),])
PSI_cands <- head(rownames(ASCA_cands[-which(rownames(ASCA_cands) %in% pathway),]),20)


cand_b <- PSI_cands[10]
cand_w <- baits[3]

cor_tw <- round(cor(X[,which(colnames(X) %in% cand_w)], X[,which(colnames(X) %in% baits[4])]), digits = 2)

cor_tb <- round(cor(X[,which(colnames(X) %in% cand_b)], X[,which(colnames(X) %in% baits[4])]), digits = 2)

cor_rw <- round(cor(res_ASCAplus_nb$residuals[,which(colnames(res_ASCAplus_nb$residuals) %in% cand_w)], res_ASCAplus_nb$residuals[,which(colnames(res_ASCAplus_nb$residuals) %in% baits[4])]), digits = 2)

cor_rb <- round(cor(res_ASCAplus_nb$residuals[,which(colnames(res_ASCAplus_nb$residuals) %in% cand_b)], res_ASCAplus_nb$residuals[,which(colnames(res_ASCAplus_nb$residuals) %in% baits[4])]), digits = 2)


within <- qqplot(cbind.data.frame("Candidate" = X[,which(colnames(X) %in% cand_w)],
                        "Bait" = X[,which(colnames(X) %in% baits[4])]),meta,
       aes(x = Bait, y = Candidate, colour = growth_condition, shape = time)) + xlab(baits[4]) + ylab(cand_w) + labs(title = paste0("r = ",cor_tw)) + theme(aspect.ratio = 1)

PSI_cands <- rownames(ASCA_cands[which(rownames(ASCA_cands) %in% psi_ref[,1]),])
PSI_cands <- head(rownames(ASCA_cands[-which(rownames(ASCA_cands) %in% pathway),]),20)


between <- qqplot(cbind.data.frame("Candidate" = X[,which(colnames(X) %in% cand_b)],
                        "Bait" = X[,which(colnames(X) %in% baits[4])]),meta,
       aes(x = Bait, y = Candidate, colour = growth_condition, shape = time)) + xlab(baits[4]) + ylab(cand_b) + labs(title = paste0("r = ",cor_tb)) + theme(aspect.ratio = 1)

# within + between + plot_layout(guides = "collect",tag_level = 'new') + plot_annotation(tag_levels = c('A')) + plot_layout(heights = unit(c(5), c("cm")), widths = c(5))




within_r <- qqplot(cbind.data.frame("Candidate" = res_ASCAplus_nb$residuals[,which(colnames(res_ASCAplus_nb$residuals) %in% cand_w)],
                        "Bait" = res_ASCAplus_nb$residuals[,which(colnames(res_ASCAplus_nb$residuals) %in% baits[4])]),meta,
       aes(x = Bait, y = Candidate, colour = growth_condition, shape = time)) + xlab(baits[4]) + ylab(cand_w)   + labs(title = paste0("r = ",cor_rw)) + theme(aspect.ratio = 1)#, alpha = (16- as.numeric(growth_condition))


between_r <- qqplot(cbind.data.frame("Candidate" = res_ASCAplus_nb$residuals[,which(colnames(res_ASCAplus_nb$residuals) %in% cand_b)],
                        "Bait" = res_ASCAplus_nb$residuals[,which(colnames(res_ASCAplus_nb$residuals) %in% baits[4])]),meta,
       aes(x = Bait, y = Candidate, colour = growth_condition, shape = time)) + xlab(baits[4]) + ylab(cand_b)  + labs(title = paste0("r = ",cor_rb)) + theme(aspect.ratio = 1)

(within + between)/(within_r + between_r) + plot_layout(guides = "collect",tag_level = 'new') + plot_annotation(tag_levels = c('A','1')) #+ plot_layout(heights = unit(c(5), c("cm")), widths = c(5))


tiff("Fig2.tif", res = 150, height = 960, width = 960)
(within + between)/(within_r + between_r) + plot_layout(guides = "collect",tag_level = 'new') + plot_annotation(tag_levels = c('A','1')) #+ plot_layout(heights = unit(c(5), c("cm")), widths = c(5))
dev.off()


```

```{r}

spikes <- sls[-which(sls %in% baits)]
resids <- res_ASCAplus_nb$residuals[,rownames(ac_top)]

spls_res <- simpls.fit(resids[,-which(colnames(resids) %in% baits)],
                          resids[,which(colnames(resids) %in% baits)], ncomp = 2)

  #####
  y <- colMeans(spls_res$Yloadings)
  U <- spls_res$projection
  UTP <- (dot(t(U),y)/dot(y,y))

  sPLS_bait_cands <- as.data.frame(UTP[order(UTP, decreasing = TRUE)])



```





```{r}
library(riceidconverter)

X <- X[,which(colnames(X) %in% colnames(resids))]

symbol <- RiceIDConvert(unique(colnames(X)),'TRANSCRIPTID',toType = 'SYMBOL')
symbol$SYMBOL <- gsub("LOC","osa:",symbol$SYMBOL)
symbol <- symbol[grep("^osa",symbol$SYMBOL),]
# KOIDs <- read.table("https://rest.kegg.jp/link/osa/ko")
path <- read.table("https://rest.kegg.jp/link/osa/pathway")
# module <- read.table("https://rest.kegg.jp/link/osa/module")


colnames(path) <- c("PATH","SYMBOL")
merged <- merge(symbol, path, by = "SYMBOL")


pnum <- as.data.frame(table(merged$PATH))
pnum <- pnum[-which(pnum$Freq < 4),]

merged <- merged[which(merged$PATH %in% pnum$Var1),]





if (!requireNamespace("KEGGREST", quietly = TRUE)) {
  install.packages("BiocManager")
  BiocManager::install("KEGGREST")
}

library(KEGGREST)


kegg_ids <- unique(gsub("path:","",merged$PATH))

# - STUPID FUNCTION BELOW -  ??

# Function to get pathway description for each ID
get_kegg_description <- function(kegg_id) {
  pathway_info <- keggGet(kegg_id)  # Query KEGG
  if (!is.null(pathway_info[[1]])) {
    return(pathway_info[[1]]$NAME)  # Extract the name/description
  } else {
    return(NA)  # Return NA if the ID is invalid
  }
}

# Apply the function to all KEGG IDs
kegg_descriptions <- sapply(kegg_ids, get_kegg_description)

# Combine IDs and descriptions into a data frame
kegg_mapping <- data.frame(
  KEGG_ID = kegg_ids,
  Description = kegg_descriptions,
  stringsAsFactors = FALSE
)

# Print the mapping
print(kegg_mapping)


kegg_DESC <- t(do.call(cbind,kegg_descriptions))






X <- X[,which(colnames(X) %in% merged$TRANSCRIPTID)]       #from ~25k to 4k IDs


cands <- sPLS_bait_cands[which(rownames(sPLS_bait_cands) %in% colnames(X)),, drop = FALSE]
cands <- cbind(rownames(cands), cands)
colnames(cands) <- c("TRANSCRIPTID","UTP")

cands <- merge(cands,merged, by = "TRANSCRIPTID")

cands <- cands[order(cands$UTP, decreasing = TRUE),]

kegg_DESC[gsub("path:","",as.data.frame(table(cands$PATH[c(nrow(cands):(nrow(cands) - 50))]))[,1]),]

kegg_DESC[gsub("path:","",as.data.frame(table(cands$PATH[1:50]))[,1]),]


resids <- resids[,which(colnames(resids) %in% colnames(X))]

res_cor <- cor(resids)
```
```{r}


diag(res_cor) <- NA

threshold <- quantile(res_cor, 0.999, na.rm = TRUE)

inds <- which(res_cor > threshold, arr.ind = TRUE)

cor_vals <- res_cor[inds]

results <- data.frame(
  Row = inds[, 1],
  Column = inds[, 2],
  Correlation = cor_vals
)

if (!is.null(rownames(res_cor))) {
  results$Variable1 <- rownames(res_cor)[results$Row]
  results$Variable2 <- colnames(res_cor)[results$Column]
}

results <- results[results$Row < results$Column, ]
results <- results[order(-results$Correlation), ]

length(unique(c(results$Variable1, results$Variable2)))

res1 <- merge(results,merged, by.x = "Variable1", by.y = "TRANSCRIPTID")
res1 <- merge(res1,merged, by.x = "Variable2", by.y = "TRANSCRIPTID")
res1 <- res1[which(res1$`PATH.x` == res1$`PATH.y`),]


cands <- cands[order(cands$UTP, decreasing = TRUE),]

kegg_DESC[gsub("path:","",as.data.frame(table(res1$`PATH.x`))[,1]),]

same_path <- as.data.frame(table(res1$`PATH.x`))
same_path <- same_path[which(same_path[,2] > 5 & same_path[,2] < 50),]

kegg_DESC[gsub("path:","",same_path[,1]),]

saveRDS(same_path, "path_example.RDS")

```


